### PR TITLE
[trainer] fix: add missing rollout dump and corrected validation logging in main_ppo_sync

### PR DESCRIPTION
**Pull request by guillemgt**
### What does this PR do?

Fixes three issues in `main_ppo_sync.py`'s validation and logging:

1. **`_dump_generations` is called but never defined.** `_validate()` calls `self._dump_generations()` (via `validation_data_dir` config) but the method doesn't exist on `PPOTrainer` — it would crash at runtime. Added the method (with numpy type serialization, which prevents crashes for int and bool types), plus `_log_rollout_data()` and the `rollout_data_dir` config hook for training rollout dumps.

1. **Validation fetches all multi-output keys instead of final-only.** With multi-output agent loops, each session produces keys `{uid}_{session_id}_{0..N}`. `_validate()` was fetching all of them for metrics, which resulted in many different `@k` metrics being logged when the number of outputs was different from session to session.
Rewards and advantages are being computed in terms of the last output in a session, so we have changed the `_validate` metrics to also be computed in terms of the last output also.

3. **`extra_fields` not fetched during validation.** Custom reward metrics in `extra_fields.reward_extra_info` were not collected, so they never appeared in val-aux logs.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: [main_ppo_sync](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+main_ppo_sync)
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

These changes are in the sync trainer's logging/validation path which requires a full distributed training setup to test end-to-end. Validated by running PPO training with multi-output agent loops and confirming:
- validate@k metrics correctly count unique sessions (not individual outputs)
- `validation_data_dir` produces JSONL dumps without crashing
- `rollout_data_dir` produces JSONL dumps sorted by uid
- `reward_extra_info` metrics appear in val-aux logs

### API and Usage Example

No changes, matches functionality in `main_ppo`

### Design & Code Changes

- `_validate()`: Filter `batch.keys` to final output per `{uid}_{session_id}` group before fetching data. Fetch `extra_fields` and extract `reward_extra_info` into `reward_extra_infos_dict`.
- `_dump_generations()`: New method (matches `ray_trainer.py`), with `json_encode_default` for numpy types.
- `_log_rollout_data()`: New method to fetch and dump training rollout data sorted by uid key.
- `fit()` training loop: Added `rollout_data_dir` config check between metrics and cleanup steps.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs). (not necessary)
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: not necessary, as there are no similar tests for `main_ppo`
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`. (not applicable)

### CODE REVIEW

File: verl/trainer/main_ppo_sync.py [hunk 1]
```diff
@@ -873,6 +899,13 @@ def _validate(self) -> dict[str, float]:
             sample_turns.extend(data.pop("num_turns").tolist())
             reward_extra_infos_dict["reward"].extend(scores)
 
+            extra_fields_list = data.pop("extra_fields", None)
+            if extra_fields_list is not None:
+                for ef in extra_fields_list.tolist():
+                    rei = ef.get("reward_extra_info", {}) if isinstance(ef, dict) else {}
+                    for key, value in rei.items():
+                        reward_extra_infos_dict[key].append(value)
```
[the remark this answers is not in the document]
Comment (guillemgt) [reply verl/trainer/main_ppo_sync.py]: Should be fixed now

**Review by wuxibin89 — APPROVED** [review]
